query-engine/query-structure/src/field_selection.rs

```diff
@@ -78,11 +75,26 @@ impl FieldSelection {
     }
 
     /// Returns all database (e.g. column or document field) names of contained fields.
-    /// Does _not_ recurse into composite selections and only iterates level fields.
+    /// Does _not_ recurse into composite selections and only iterates top level fields.
+    /// Returns db aliases for virtual fields grouped into objects in the query separately,
+    /// representing results of queries that do not load relations using JOINs.
     pub fn db_names(&self) -> impl Iterator<Item = String> + '_ {
         self.selections.iter().map(|f| f.db_name().into_owned())
     }
 
+    /// Returns all database (e.g. column or document field) names of contained fields. Does not
+    /// recurse into composite selections and only iterates top level fields. Also does not recurse
+    /// into the grouped containers for virtual fields, like `_count`. The names returned by this
+    /// method correspond to the results of queries that use JSON objects to represent joined
+    /// relations and relation aggregations.
+    pub fn db_names_grouping_virtuals(&self) -> impl Iterator<Item = String> + '_ {
+        self.selections
+            .iter()
+            .map(|f| f.db_name_grouping_virtuals())
+            .unique()
+            .map(Cow::into_owned)
+    }
```

Comment:
I would prefer the grouping to be done _once_ at the `selections()` level instead of having to `unique` on every method that needs this grouping.

(pseudo code)
```rust
pub fn selections_grouped_virtuals(&self) -> impl Iterator<Item = &SelectedField> + '_ {
  self.selections().<group_by_virtuals...>
}

pub fn db_names_grouping_virtuals(&self) -> impl Iterator<Item = String> + '_ {
    self.selections_grouped_virtuals()
        .map(|f| f.db_name_grouping_virtuals())
        .map(Cow::into_owned)
}

pub fn type_identifiers_with_arities_grouping_virtuals(&self) -> Vec<(TypeIdentifier, FieldArity)> {
    self.selections_grouped_virtuals()
        .filter_map(|vs| vs.type_identifier_with_arity())
        .collect()
}
```

Reply:
The reason I didn't do it like this is that it's not clear what a grouped `SelectedField` should be. If it returns the first field in each group, then it just loses information and violates the expectations that the caller would have based on the signature and types, and wouldn't really be useful outside of being an implementation detail of `db_names_grouping_virtuals` and `type_identifiers_with_arities_grouping_virtuals`. If it actually groupspartitions fields, iterating over vectors of grouped fields, then we pay that cost even though we don't actually need it.

But now that I think about it, a method that returns an iterator over first fields of each group would actually be fine as private method?

Reply:
done